storage/src/rocksdb_storage/storage_context/context_tx.rs

```diff
@@ -33,39 +33,111 @@ use grovedb_costs::{
+        },
+        storage_context::context_no_tx::PrefixedSecondaryRocksDbStorageContext,
+    },
     RawIterator, StorageBatch, StorageContext,
 };
 
 /// Storage context with a prefix applied to be used in a subtree to be used in
 /// transaction.
-pub struct PrefixedRocksDbTransactionContext<'db> {
-    storage: &'db Db,
+pub struct PrefixedPrimaryRocksDbTransactionContext<'db> {
+    storage: &'db OptimisticTransactionDB,
     transaction: &'db Tx<'db>,
     prefix: SubtreePrefix,
     batch: Option<&'db StorageBatch>,
 }
 
+/// Prefixed rocks db transaction context
+pub enum PrefixedRocksDbTransactionContext<'db> {
+    /// Primary storage context
+    Primary(PrefixedPrimaryRocksDbTransactionContext<'db>),
+    /// Secondary storage context
+    Secondary(PrefixedSecondaryRocksDbStorageContext<'db>),
+}
+
+macro_rules! call_with_storage_prefix_and_batch {
+    ($self:ident, $storage:ident, $prefix:ident, $batch:ident, $code:block) => {
+        match $self {
+            PrefixedRocksDbTransactionContext::Primary(context) => {
+                let $storage = context.storage;
+                let $prefix = &context.prefix;
+                let $batch = context.batch;
+
+                $code
+            }
+            PrefixedRocksDbTransactionContext::Secondary(context) => {
+                let $storage = context.storage;
+                let $prefix = &context.prefix;
+                let $batch = context.batch;
+
+                $code
+            }
+        }
+    };
+}
+
+macro_rules! call_with_storage_or_transaction_prefix_and_batch {
+    ($self:ident, $storage:ident, $prefix:ident, $batch:ident, $code:expr) => {
+        match $self {
+            PrefixedRocksDbTransactionContext::Primary(context) => {
+                let $storage = context.transaction;
+                let $prefix = &context.prefix;
+                let $batch = context.batch;
+
+                $code
+            }
+            PrefixedRocksDbTransactionContext::Secondary(context) => {
+                let $storage = context.storage;
+                let $prefix = &context.prefix;
+                let $batch = context.batch;
+
+                $code
+            }
+        }
+    };
+}
+
 impl<'db> PrefixedRocksDbTransactionContext<'db> {
-    /// Create a new prefixed transaction context instance
-    pub fn new(
-        storage: &'db Db,
+    /// Create a new prefixed context instance
+    pub fn new_primary(
+        storage: &'db OptimisticTransactionDB,
         transaction: &'db Tx<'db>,
         prefix: SubtreePrefix,
         batch: Option<&'db StorageBatch>,
     ) -> Self {
-        PrefixedRocksDbTransactionContext {
+        let context = PrefixedPrimaryRocksDbTransactionContext {
             storage,
             transaction,
             prefix,
             batch,
-        }
+        };
+
+        Self::Primary(context)
+    }
+
+    /// Create a new prefixed context instance
+    pub fn new_secondary(
```

Comment:
it's `context_tx` file, we're dealing with transactions there and this one receives none as it should

so I suggest to have no such context at all because we have no such case for secondary